### ISSUE TITLE
Question about logprobs_from_logits...

### ISSUE THREAD

**Issue by wulaoshi** [huggingface/trl#1061]
This works great!

```
def logprobs_from_logits(logits, labels, gather=True):
    logp = F.log_softmax(logits, dim=2)

    if not gather:
        return logp
    logpy = torch.gather(logp, 2, labels.unsqueeze(2)).squeeze(-1)
    return logpy
```

I had a question while reading the code: I don't understand why in the code it is doing the log_softmax operation before the gather operation. If we want to calculate the kl_loss of the model and model_ref, we should perform the gather operation before the log_softmax operation. Can you tell me the execution logic here? Thanks!

**Comment by lvwerra** [comment]
You need to calculate the softmax over all possible tokens at each position to get a probability (or log probability) and then in the gather you only select the probabilities of the decoded tokens. Does that make sense?

**Comment by wulaoshi** [comment]
> You need to calculate the softmax over all possible tokens at each position to get a probability (or log probability) and then in the gather you only select the probabilities of the decoded tokens. Does that make sense?

Aha, I see now! So, first, we compute the softmax over all possible tokens at each position to obtain their probabilities (or log probabilities). Then, we use the gather function to extract the probabilities of only the decoded tokens. That makes perfect sense! Thank you so much for the clear explanation, I really appreciate your help!

**Comment by wulaoshi** [comment]
> You need to calculate the softmax over all possible tokens at each position to get a probability (or log probability) and then in the gather you only select the probabilities of the decoded tokens. Does that make sense?

Aha, I see now! So, first, we compute the softmax over all possible tokens at each position to obtain their probabilities (or log probabilities). Then, we use the gather function to extract the probabilities of only the decoded tokens. That makes perfect sense! Thank you so much for the clear explanation, I really appreciate your help!
